### PR TITLE
Improvement: Allow subscriptions for complete sitemaps (not limited to a single page)

### PR DESCRIPTION
**Pull request by TAKeanice**
The changes in this PR allow to leave out the pageId in a request for updates from a subscription. The motivation for this was discussed in https://github.com/openhab/openhab-core/issues/3650

Leaving out the pageId in a subscription request makes the event API send events for updates of _all_ widgets within a sitemap. This is accomplished by walking the sitemap tree and collecting all widgets in it when the subscription is created. Previously, only the widgets on the given page were collected at that occasion.

In the issue above, performance issues were discussed. I do not know how to test for them, and would happily accept help with that matter.

### CODE REVIEW

**Comment by lolodomo** [comment]
I will have a look to your proposal. But as this is something we are not recommending at all, first can you confirm that this will not become the default behaviour for existing UIs ? And I would like something added in the javadoc explaining that this is a possible not recommended usage.
Any official openHAB should not use that. This could lead to a massive network traffic between server and browser (SSE events).

**Comment by TAKeanice** [comment]
@lolodomo thank you for giving this PR some support. I split up the methods to create a sitemap-wide subscription and a normal subscription endpoint, and commented the ones that should not be used. There should be no possibility to unconsciously subscribe to the whole sitemap.

The things I did not change apparently contained some nullability issues with the pageId. I did not try to resolve them to not break legacy behaviour.

**Comment by TAKeanice** [comment]
@lolodomo the compliance with your request to make sure not the whole sitemap is listened to by default is now also ensured by a regression test. I don´t see any reason why this should not be merged.

**Comment by TAKeanice** [comment]
@lolodomo I know everyone is busy and stuff, but now that it´s spring and everything is getting back to life, would you consider reviving your interest in this PR, too?

File: bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/SitemapSubscriptionService.java
Comment (J-N-K) [inline bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/SitemapSubscriptionService.java:356]:
```suggestion
    private String getScopeIdentifier(String sitemapName, @Nullable String pageId) {
        return pageId == null ? sitemapName : sitemapName + SITEMAP_PAGE_SEPARATOR + pageId;
    }
```

I think you can also collapse both versions of `addCallbackToListener` in one method, calling either `collectAllWidgetsInSitemap` or `collectWidgetsOnPage` (depending on `pageId == null`).

File: bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/internal/SitemapResource.java
Comment (J-N-K) [inline bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/internal/SitemapResource.java:279]: I would prefer something like `*`, because `wholeSitemap` is also a valid page id (at least it matchs the regexp).

File: bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/SitemapSubscriptionService.java
Comment (J-N-K) [inline bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/SitemapSubscriptionService.java:236]: I don't think you need this method if you allow `pageId` to be `null` and act accordingly. This also simplifies code in the `SitemapResource`

**Review by J-N-K — CHANGES_REQUESTED** [review]
I think you can simplify the code very much by allowig `pageId` to be null (and consider that as "whole sitemap"). You can remove nearly all added methods in this case.

**Comment by TAKeanice** [comment]
@J-N-K I the request by @lolodomo was specifically not to do that, since it could accident is be used in some official implementation. My original implementation was like you suggested. If you want, I revert my changes to get that version back. Expecting your final decision :)

**Comment by J-N-K** [comment]
I can't find that request in this PR, can you point me to the comment?

**Comment by TAKeanice** [comment]
> the comment

https://github.com/openhab/openhab-core/pull/3652#issuecomment-1773755492_

Have I misinterpreted the request to make sure it does not get used accidentally?

**Comment by J-N-K** [comment]
IMO yes, my understanding is that a comment should state "whole sitemap is not recommended" (maybe also in the API documentation, not only in JavaDoc) and that a subscription with page should never result in a subscription to the whole sitemap.

**Comment by TAKeanice** [comment]
I will make some simplifications to the private methods, that should reduce code duplication.

**Comment by lolodomo** [comment]
> IMO yes, my understanding is that a comment should state "whole sitemap is not recommended

Yes, that was my request.
Also as a consequence that it should not become the default behaviour.

**Comment by TAKeanice** [comment]
The reason I split the methods is that the method `getSitemapEvents` https://github.com/openhab/openhab-core/blob/81f2bd9366d3f937da892ac704671207fce5e10c/bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/internal/SitemapResource.java#L346-L349 allows null as pageId. If I unsplit the whole implementation by allowing null pageIds throughout, omitting the pageId when calling that method would result in accidentally using sitemap-wide subscriptions. That was the nullability issue I referred to here: https://github.com/openhab/openhab-core/pull/3652#issuecomment-1807185491

**Comment by J-N-K** [comment]
It does not allow `null` as `pageId`:

```
 if (sitemapname != null && pageId != null) {
            subscriptions.setPageId(subscriptionId, sitemapname, pageId);
        }
```

My proposal was to use `null` in Java code and `*` as `pageId` for "whole sitemap". `*` is not a valid `pageId`, so there can be no collisions. You can adjust the regexp so that it accepts either `*` or a valid `pageId`.

**Comment by TAKeanice** [comment]
I improved the code as suggested

File: bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/SitemapSubscriptionService.java
Comment (J-N-K) [inline bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/SitemapSubscriptionService.java:320]:
```suggestion
        return pageId == null ? sitemapName : sitemapName + SITEMAP_PAGE_SEPARATOR + pageId;
```

File: bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/internal/SitemapResource.java
Comment (J-N-K) [inline bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/internal/SitemapResource.java:285]:
```suggestion
    public Response getSitemapData(@Context HttpHeaders headers,
```

**Review by J-N-K — CHANGES_REQUESTED** [review]
Otherwise LGTM, @lolodomo can you have a look or test this before merge?

**Comment by J-N-K** [comment]
There seems to be an issue with spotless

**Comment by TAKeanice** [comment]
> There seems to be an issue with spotless

yes, it does not like formatting in conditions :D Fixed that.

File: bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/internal/SitemapResource.java
Comment (lolodomo) [inline bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/internal/SitemapResource.java:391]: Please update the description.
I have even the felling this response can not occur; if that is true, rather remove this line.

**Review by lolodomo — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/SitemapSubscriptionService.java
Comment (lolodomo) [inline bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/SitemapSubscriptionService.java:237]: Typo: page **id** is null

File: bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/SitemapSubscriptionService.java
Comment (lolodomo) [inline bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/SitemapSubscriptionService.java:76]: Minor but rather private

**Review by lolodomo — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/internal/SitemapResource.java
Comment (lolodomo) [inline bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/internal/SitemapResource.java:306]: Why do you need timeout information in the response ?
It was not needed in the current implementation.

Comment (TAKeanice) [reply bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/internal/SitemapResource.java:306]: Right, some of these changes might have crept in because I started development quite some time ago when the code still was like that. Same with the nullability problems. I will fix it this evening.

Comment (TAKeanice) [reply bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/internal/SitemapResource.java:306]: I remember why the timeout field is necessary. It was meant analogous to the timeout in PageDTO, but I can just use the field timeout in the homepage of the sitemap.

Comment (TAKeanice) [reply bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/internal/SitemapResource.java:306]: Timeout parameter removed from SitemapDTO, but still needed in getSitemapBean to set on homepage PageDTO. The parameter is necessary to tell the caller whether a change has occurred or just a timeout when long-polling for events

File: bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/SitemapSubscriptionService.java
Comment (TAKeanice) [reply bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/SitemapSubscriptionService.java]: Is now private

File: bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/SitemapSubscriptionService.java
Comment (TAKeanice) [reply bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/SitemapSubscriptionService.java]: wrote pageId to use the same spelling as the parameter

File: bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/internal/SitemapResource.java
Comment (TAKeanice) [reply bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/internal/SitemapResource.java]: The method `getSitemapEvents` returns a `BAD_REQUEST` when there is either no sitemap or no pageId (when trying to subscribe to a page only) or when either of them could not be set as location for the subscription

File: bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/internal/SitemapResource.java
Comment (J-N-K) [inline bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/internal/SitemapResource.java:408]:
```suggestion
            @ApiResponse(responseCode = "400", description = "Missing sitemap or page parameter, or page not linked successfully to the subscription."),
```

**Review by J-N-K — CHANGES_REQUESTED** [review]
Otherwise LGTM, maybe @lolodomo can also approve here.

File: bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/internal/SitemapResource.java
Comment (lolodomo) [inline bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/internal/SitemapResource.java:431]: I believe some parenthesis are missing around the second line. Please confirm.

File: bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/SitemapSubscriptionService.java
Comment (lolodomo) [inline bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/SitemapSubscriptionService.java:237]: Very minor but you still kept the doubled is => is is null

**Review by lolodomo — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/internal/SitemapResource.java
Comment (J-N-K) [reply bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/internal/SitemapResource.java:431]: good catch

Comment (TAKeanice) [reply bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/internal/SitemapResource.java:431]: No, because && has higher operator precedence.

Comment (TAKeanice) [reply bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/internal/SitemapResource.java:431]: The condition says "we need to have successfully subscribed to a sitemap and, if we do not want to subscribe to a whole sitemap, we need to have successfully subscribed to a page"

File: bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/internal/SitemapResource.java
Comment (lolodomo) [inline bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/internal/SitemapResource.java:809]: This will be interesting that you add a DEBUG log just after this line showing how many widgets you have and how many items it resulted.

File: bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/internal/SitemapResource.java
Comment (lolodomo) [reply bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/internal/SitemapResource.java:431]: Even if the code is good due to precedence of such operator over another, I will prefer you add parenthesis as precedence between && and || is not something very intuitive.

File: bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/internal/SitemapResource.java
Comment (lolodomo) [reply bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/internal/SitemapResource.java:809]: In a common user situation (user defining only one sitemap), it should lead to have more or less all items.

Comment (TAKeanice) [reply bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/internal/SitemapResource.java:780]: Good idea, I added that but in the method in SitemapSubscriptionService and deduplicated the code from this method.

File: bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/internal/SitemapResource.java
Comment (TAKeanice) [reply bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/internal/SitemapResource.java:431]: Matter of taste, I always prefer less boilerplate but will follow your preference here as you´re the maintainers :) To explain my preference, I think in complicated boolean statements, too many parenthesis in a row confuse new developers, as they have to count the parenthesis to know what they enclose. I think, although not coming from a Python background, that smart indentations help understanding much better. For that statement for example I would prefer the following, but I believe spotless prohibits that:

```Java
if (subscriptions.getSitemapName(subscriptionId) == null
     || subscriptions.getPageId(subscriptionId) == null 
        && !subscribeToWholeSitemap)
```

File: bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/internal/SitemapResource.java
Comment (TAKeanice) [reply bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/internal/SitemapResource.java:780]: Sorry, I got that wrong, added the log statement into waitForChanges as well.

File: bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/SitemapSubscriptionService.java
Comment (lolodomo) [inline bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/SitemapSubscriptionService.java:289]: In the current implementation, `pageWidget`was not considered by `blockUnlessChangeOccurs`.

I am not sure if it was an error or it was not an error what could be the impact of considering it now.
WDYT ?

**Comment by J-N-K** [comment]
There is an issue with the tests, can you fix that? I'll nmerge afterwards. Thanks for your patience.

**Comment by lolodomo** [comment]
I have not checked but I will not be surprised that tests are failing due to the change I am commenting in my last message.

**Comment by TAKeanice** [comment]
@J-N-K and @lolodomo I think the tests fail due to my code deduplication, I'll fix that today. It was a pleasure working on finalizing this the last couple of days, thank you for that!

File: bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/SitemapSubscriptionService.java
Comment (TAKeanice) [reply bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/SitemapSubscriptionService.java:289]: I think both implementations for collecting widgets are meant to do the same and omitting the page in one of them was the typical consequence of code duplication. I will check what that causes with the tests though.

**Comment by TAKeanice** [comment]
The test fails because the subscriptionService is just a mock, I need to properly instantiate it and mock its itemUIRegistry in the test.

**Comment by J-N-K** [comment]
You can also keep the mocked `subscriptionService` and return another mock, depending on the called method (using `when`).

**Comment by TAKeanice** [comment]
> You can also keep the mocked `subscriptionService`

but that would defeat the purpose of the test, since an important part of the logic resides in collectWidgets

**Comment by J-N-K** [comment]
I didn't look at the test, it was more a general comment.

File: bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/SitemapSubscriptionService.java
Comment (lolodomo) [reply bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/SitemapSubscriptionService.java:289]: I just hope it was not done deliberately.

Comment (TAKeanice) [reply bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/SitemapSubscriptionService.java:289]: I will take a look at the history of the code before resolving this conversation

Comment (lolodomo) [reply bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/SitemapSubscriptionService.java:289]: Is `blockUntilChangeOccurs` only used when long polling mode is used ? Sorry, I do not remember.

Comment (TAKeanice) [reply bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/SitemapSubscriptionService.java:289]: Yes

Comment (lolodomo) [reply bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/SitemapSubscriptionService.java:289]: In this case, this will not impact Basic UI when used with any modern browser (meaning supporting SSE).
But not sure about Android app and iOS app, even if I know that the Android app is also using a SSE connection.

By the way, as previously said, the change probably makes sense.

**Review by J-N-K — APPROVED** [review]
Thanks!

File: bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/SitemapSubscriptionService.java
Comment (TAKeanice) [reply bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/SitemapSubscriptionService.java:289]: Seems like both places have been there since the code was added to the repository in 2019 (a37cceab), so I can´t judge. But the effect will only be a little more frequent updates, that occur also when the parent page itself changes, so I agree that it will probably not cause any trouble.

**Comment by TAKeanice** [comment]
@J-N-K the test issues have been resolved.

**Comment by lolodomo** [comment]
By renaming methods, you broke UIs using them :(
Basic UI is no more compiling, I am going to fix that.

**Comment by lolodomo** [comment]
I search in GitHub and apparently Basic UI was the unique app calling this method.

**Pull request merged**
